[BD-14] [SE-2937] Authoring page

Kelketek opened the pull request:
This PR introduces the new 'Authoring Page' which is intended to replace the current Library Detail view (pending a follow-up PR to be made by @arbrandes containing complex library capabilities).

The PR has a few tricks up its sleeve-- the first is that it generalizes the block editing slice to allow for multiple blocks to be tracked, thus permitting us to view the previews of all the blocks in a library at once.

The second is that it handles showing and hiding of previews, and saving the user's preference to localStorage.

The third is that it refines some of the testing tools that were made previously to handle cases that came out of testing this PR. One new feature is the ability to spy on dispatch calls, and another is mocks for local storage. The testing environment now better mirrors real world use, allowing for greater fidelity in tests.

One more thing this new page offers is that it is rewritten using Paragon and does not share its markup with the studio. This has allowed much more succinct templates.

**Discussions**: https://docs.google.com/document/d/1hRrSmWWlGHhi-bh9AIlpSuLtJFnn2GpavNe8OQxthbE/

**Dependencies**: None

**Screenshots**:

![image](https://user-images.githubusercontent.com/1099483/95004455-cefcf780-05b0-11eb-817c-1c7643ad5a8c.png)

![image](https://user-images.githubusercontent.com/1099483/95004462-ea680280-05b0-11eb-93b0-6fd9fbe1e992.png)

WIth previews hidden:

![image](https://user-images.githubusercontent.com/1099483/95004469-05d30d80-05b1-11eb-8402-b9131dbabd3d.png)


**Merge deadline**: Sooner is better than later.

**Testing instructions**:

1. Create a new Video/Problem library, or visit an existing one.
2. Use the 'Add Component' button, which should be renamed 'Add Video' or 'Add Problem' depending on the library type you're working with. Verify a new block has been created and that its preview loads. (It may be slow in a dev environment).
3. Add another component.
4. Toggle show and hide previews. Verify that the setting is saved on refresh.
5. Delete components.
6. Visit a complex library and verify it goes to the old LibraryPage instead of the LibraryAuthoringPage for now.
7. Publish your changes.
8. Change something.
9. Revert your changes. (It may take a moment)

**Author notes and concerns**:

1. The studio has a 'copy' button which is not implemented here. I attempted to implement it, but could not get copying to work using the existing endpoint used by old libraries, nor did I see an obvious endpoint in the new API for doing so. I imagine we'll want this feature in the future, so unless I've missed something, this will probably need to be a follow-up task.
2. The 'edit' button exists but can't be rendered currently because it only displays for blocks that can be edited, and neither the Problem nor the Video blocks can be.
3. As an addendum to 2, the editing functionality would currently go to the existing block edit page anyway-- I didn't flesh it out since we can't use it at present. Having settings appear in a modal will need to be a future task.

**Reviewers**
- [ ] @arbrandes 
- [ ] @kdmccormick

edx-webhook commented:
Thanks for the pull request, @Kelketek! I've created [BLENDED-611](https://openedx.atlassian.net/browse/BLENDED-611) to keep track of it in Jira.   More details are on the [BD-14](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1545011241/BD-14+Blockstore+Powered+Content+Libraries+Taxonomies) project page.


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

arbrandes commented on src/library-authoring/library-detail/LibraryAuthoringPage.jsx:
```diff
@@ -0,0 +1,435 @@
+      blockId: block.id,
+    });
+  }, []);
+  useEffect(() => {
```
TIL about `useEffect`.  +1

arbrandes commented on src/library-authoring/library-detail/LibraryAuthoringPage.jsx:
```diff
@@ -0,0 +1,435 @@
+  // will show the student view, and navigating back will show the edit view in the block list. React is smart enough
+  // to guess these iframes are the same between routes and will try to preserve rather than rerender, but that works
+  // against us here. Setting an explicit key prevents it from matching the two.
+  const previewKey = useMemo(() => `${uuid4()}`, [block.id]);
```
This explains why I could not get iframes to refresh - sometimes.  TIL how to fix it.  +1 again.

arbrandes reviewed (changes requested):
@Kelketek, it was a pleasure reading the code (learned quite a few new tricks, thanks!), and everything I tested works, with one exception which I'll list below.  Congrats!

Changes I think are a good idea to include in this PR:

* Looks like there's a state-cleanup issue: if I start from scratch and create a new Video library, everything works.  If I then create a Complex library, and either navigate back to a Video library or create a new one, I can no longer add video blocks.  The button says "Add component", and the library title in the header still show the Complex library info.
* The "Add video" button on the top right doesn't seem to work.
* There are a few of styling/markup issues.  See screenshot below.

![Screenshot_2020-10-05_13-41-04](https://user-images.githubusercontent.com/759355/95108363-958fcd80-0711-11eb-894f-a41c4e54be68.png)

1. The "Library Name" font does not match what's used in Studio (and the old library page).  Also, a little more vertical spacing is required between heading and subheading.
2. Buttons should be larger on the top right.
3. Can we get some spacing between the video entries, particularly when previews are collapsed?
4. Can the "Add component/video/problem" button be put into the white container above, with appropriate spacing?  This will help make the authoring page less forlorn when empty.  The button can be made a larger, as well.  (No need to make it square, for now, though there's a good chance we'll get instructions to do so in the future.)

Kelketek commented:
@arbrandes I've addressed the visual notes. The state change issue is something I think we should circle back to-- it appears to be due to the fact that several other apps use that libraryDetailSelector to get their library info, but after having gotten it, don't update the library afterward. I think it's worth its own ticket to clean up those issues or else refactor how we store the focused library's info.

arbrandes commented:
@Kelketek, thanks for addressing the issues quickly!

> several other apps use that libraryDetailSelector to get their library info, but after having gotten it, don't update the library afterward

Do I understand correctly that the problem is that the new authoring page expects them to clean up after themselves, whereas prior to this changeset the general expectation was different?  If so, it doesn't make sense to merge this without a proper refactor, wouldn't you agree?  As it stands, even if adding just video blocks I run into the bug.

Also, I tested again and still can't add a block using the top right button.

And bear with me, as I found a couple more styling issues: note how the legacy library page organizes the sidebar on the right.  The buttons are above the heading line but the content is below it, including some help text above the publish box:

![Screenshot_2020-10-05_16-26-32](https://user-images.githubusercontent.com/759355/95123992-2f16a980-0729-11eb-9421-3ee792ca51b4.png)

Mind doing the same, here?  Also, the subheadings on the left need a little more vertical spacing.

![Screenshot_2020-10-05_16-24-59](https://user-images.githubusercontent.com/759355/95123995-2faf4000-0729-11eb-99e9-3d2225f7beb5.png)

Thanks, and sorry for all the CSS nit-picking.

Kelketek commented:
@arbrandes Turns out there were two bugs and one of them was mine and the other was that cleanup issue. They're both fixed. I also fixed the button (missed that note before) and re-arranged the page to better mirror the original. I've also added instructional text.

arbrandes commented on src/library-authoring/library-detail/LibraryAuthoringPage.jsx:
```diff
@@ -357,8 +369,8 @@ export const LibraryAuthoringPageContainerBase = ({
     const { libraryId } = props.match.params;
-    if (library === null) {
-      props.fetchLibraryDetail({ libraryId });
+    if (!library || (library && library.id !== libraryId)) {
```
Ok, I see what the problem was, and I agree you nailed it.  Thanks!

arbrandes commented on src/library-authoring/library-detail/messages.js:
```diff
@@ -151,6 +151,37 @@ const messages = defineMessages({
+      + 'settings for each Randomized Content Block, select this library as the source library, and specify the number '
+      + 'of problems to be randomly selected and displayed to each student.',
+    description: 'The first (and presently only) paragraph of instructional text about how to use library content.',
+  },
```
Thanks for adding in the sample help text.

arbrandes reviewed (approved):
@Kelketek , thanks for bearing with me!  This looks and works very well.  :+1:

- [x] I tested this as instructed.
- [x] I read through the code.
- [x] I checked for accessibility issues.
- [x] Includes sufficient inline documentation.

Kelketek commented:
@kdmccormick This is ready for you.

kdmccormick commented:
Ack'ing; thank you @Kelketek !

kdmccormick commented:
I was hoping to review this by today, but I've been consumed this week by [TNL-7425](https://openedx.atlassian.net/browse/TNL-7425).

When you get a chance @Kelketek , could you do a rebase? I'm hoping to review this in the first half of next week.

Kelketek commented:
@kdmccormick Rebased.

kdmccormick commented on src/library-authoring/library-detail/LibraryAuthoringPage.jsx:
```diff
@@ -0,0 +1,478 @@
+
+  const newBlock = () => {
+    setSending(true);
+    props.createLibraryBlock({
+      data: {
+        libraryId: library.id,
+        block_type: library.type,
+        definition_id: `${uuid4()}`,
+      },
+    }).finally(() => setSending(false));
```
I think you also need to pass `libraryId` in the outer dictionary. As-is, this leads to `POST /api/libraries/v2/undefined/blocks` for me.

kdmccormick reviewed (changes requested):
Walked through the testing instructions. Great work overall. Two issues:
* "Add Problem/Video" was broken (see comment), but I was able to fix it locally. Probably just the result of a rebase.
* The video previews are comically jittery for me (on hover in Chrome, and at all times in FF). I remember @arbrandes mentioning that the old `author_view`s might cause these kind of rendering problems. If it's not something you think you can fix quickly and without touching edx-platform, just drop a note in Known Issues.

About to do a skim-through of the code, but I wanted to give this feedback ASAP.

kdmccormick commented on src/library-authoring/library-detail/specs/LibraryAuthoringPage.spec.jsx:
```diff
@@ -0,0 +1,74 @@
+} from '../../common/specs/factories';
+import LibraryAuthoringPageContainer from '../LibraryAuthoringPage';
+import { STORE_NAMES } from '../../common/data';
+
```
Could you add a test for the `Add {blocktype}` button?

Kelketek replied:
This test has been written, but it's in the complex authoring PR, where I added a bunch more helpers to allow more comprehensive and useful introspection during the tests.

kdmccormick reviewed (changes requested):
Code looks good to me; I only found one more thing to comment on.

Once you've resolved these comments, I'm good to merge this. Thanks for bearing with the slow review cycle again.

Kelketek commented:
@kdmccormick Fixed that thunk call, added a note in the README. I don't think that's going to be an easy fix. We might be able to sort of ration out how many calls are made to the backend at once so it's not so unwieldy, but there's definitely a lot that goes into loading each of those blocks.
